File: backend/app/repositories/StudentRepository.scala
package repositories

import com.datastax.oss.driver.api.core.cql.Row
import models._
import services.CassandraService
import javax.inject._
import scala.concurrent.{ExecutionContext, Future}
import scala.jdk.CollectionConverters._
import java.time.Instant
import java.util.UUID
import play.api.Logger

@Singleton
class StudentRepository @Inject()(
  cassandra: CassandraService
)(implicit ec: ExecutionContext) {
  
  private val logger = Logger(this.getClass)
  private val keyspace = "subjectplanning"
  
  def findByEmail(email: String): Future[Option[Student]] = {
    val query = s"""
      SELECT * FROM $keyspace.students 
      WHERE email = '$email' 
      ALLOW FILTERING
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToStudent(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding student by email: ${e.getMessage}", e)
        None
    }
  }
  
  def findById(id: UUID): Future[Option[Student]] = {
    val query = s"""
      SELECT * FROM $keyspace.students 
      WHERE id = $id
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToStudent(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding student by id: ${e.getMessage}", e)
        None
    }
  }
  
  def findByStudentId(studentId: String): Future[Option[Student]] = {
    val query = s"""
      SELECT * FROM $keyspace.students 
      WHERE student_id = '$studentId' 
      ALLOW FILTERING
    """
    
    cassandra.executeAsync(query).map { resultSet =>
      val row = resultSet.one()
      if (row != null) Some(mapRowToStudent(row))
      else None
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding student by student_id: ${e.getMessage}", e)
        None
    }
  }
  
  def create(student: Student): Future[Student] = {
    val query = s"""
      INSERT INTO $keyspace.students 
      (id, student_id, name, email, password_hash, gpa, semester, created_at, updated_at)
      VALUES (
        ${student.id},
        '${student.studentId}',
        '${student.name}',
        '${student.email}',
        '${student.passwordHash}',
        ${student.gpa},
        ${student.semester},
        '${student.createdAt}',
        '${student.updatedAt}'
      )
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Created student: ${student.email}")
      student
    }.recover {
      case e: Exception =>
        logger.error(s"Error creating student: ${e.getMessage}", e)
        throw e
    }
  }
  
  def update(student: Student): Future[Student] = {
    val query = s"""
      UPDATE $keyspace.students 
      SET name = '${student.name}',
          gpa = ${student.gpa},
          semester = ${student.semester},
          updated_at = '${Instant.now()}'
      WHERE id = ${student.id}
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Updated student: ${student.id}")
      student.copy(updatedAt = Instant.now())
    }.recover {
      case e: Exception =>
        logger.error(s"Error updating student: ${e.getMessage}", e)
        throw e
    }
  }
  
  def delete(id: UUID): Future[Boolean] = {
    val query = s"""
      DELETE FROM $keyspace.students 
      WHERE id = $id
    """
    
    cassandra.executeAsync(query).map { _ =>
      logger.info(s"Deleted student: $id")
      true
    }.recover {
      case e: Exception =>
        logger.error(s"Error deleting student: ${e.getMessage}", e)
        false
    }
  }
  
  def findAll(): Future[List[Student]] = {
    val query = s"SELECT * FROM $keyspace.students"
    
    cassandra.executeAsync(query).map { resultSet =>
      resultSet.currentPage().asScala.map(mapRowToStudent).toList
    }.recover {
      case e: Exception =>
        logger.error(s"Error finding all students: ${e.getMessage}", e)
        List.empty
    }
  }
  
  private def mapRowToStudent(row: Row): Student = {
    Student(
      id = row.getUuid("id"),
      studentId = row.getString("student_id"),
      name = row.getString("name"),
      email = row.getString("email"),
      passwordHash = row.getString("password_hash"),
      gpa = row.getDouble("gpa"),
      semester = row.getInt("semester"),
      createdAt = row.getInstant("created_at"),
      updatedAt = row.getInstant("updated_at")
    )
  }
}

File: backend-python/app/repositories/student_repository.py
from typing import Optional, List
from uuid import UUID
from datetime import datetime
import logging
from app.models import Student, StudentCreate
from app.services.cassandra_service import cassandra_service
from passlib.hash import bcrypt

logger = logging.getLogger(__name__)


class StudentRepository:
    def __init__(self):
        self.session = cassandra_service.get_session()
        self.keyspace = "subjectplanning"
    
    def find_by_email(self, email: str) -> Optional[Student]:
        """Find student by email"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.students 
                WHERE email = %s 
                ALLOW FILTERING
            """
            result = self.session.execute(query, (email,))
            row = result.one()
            
            if row:
                return self._map_row_to_student(row)
            return None
        except Exception as e:
            logger.error(f"Error finding student by email: {str(e)}")
            return None
    
    def find_by_id(self, student_id: UUID) -> Optional[Student]:
        """Find student by ID"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.students 
                WHERE id = %s
            """
            result = self.session.execute(query, (student_id,))
            row = result.one()
            
            if row:
                return self._map_row_to_student(row)
            return None
        except Exception as e:
            logger.error(f"Error finding student by id: {str(e)}")
            return None
    
    def find_by_student_id(self, student_id: str) -> Optional[Student]:
        """Find student by student ID number"""
        try:
            query = f"""
                SELECT * FROM {self.keyspace}.students 
                WHERE student_id = %s 
                ALLOW FILTERING
            """
            result = self.session.execute(query, (student_id,))
            row = result.one()
            
            if row:
                return self._map_row_to_student(row)
            return None
        except Exception as e:
            logger.error(f"Error finding student by student_id: {str(e)}")
            return None
    
    def create(self, student_data: StudentCreate) -> Student:
        """Create a new student"""
        try:
            # Hash password
            password_hash = bcrypt.hash(student_data.password)
            
            student = Student(
                student_id=student_data.student_id,
                name=student_data.name,
                email=student_data.email,
                password_hash=password_hash,
                gpa=student_data.gpa,
                semester=student_data.semester
            )
            
            query = f"""
                INSERT INTO {self.keyspace}.students 
                (id, student_id, name, email, password_hash, gpa, semester, created_at, updated_at)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            """
            
            self.session.execute(query, (
                student.id,
                student.student_id,
                student.name,
                student.email,
                student.password_hash,
                student.gpa,
                student.semester,
                student.created_at,
                student.updated_at
            ))
            
            logger.info(f"Created student: {student.email}")
            return student
        except Exception as e:
            logger.error(f"Error creating student: {str(e)}")
            raise
    
    def update(self, student: Student) -> Student:
        """Update an existing student"""
        try:
            student.updated_at = datetime.utcnow()
            
            query = f"""
                UPDATE {self.keyspace}.students 
                SET name = %s, gpa = %s, semester = %s, updated_at = %s
                WHERE id = %s
            """
            
            self.session.execute(query, (
                student.name,
                student.gpa,
                student.semester,
                student.updated_at,
                student.id
            ))
            
            logger.info(f"Updated student: {student.id}")
            return student
        except Exception as e:
            logger.error(f"Error updating student: {str(e)}")
            raise
    
    def delete(self, student_id: UUID) -> bool:
        """Delete a student"""
        try:
            query = f"""
                DELETE FROM {self.keyspace}.students 
                WHERE id = %s
            """
            self.session.execute(query, (student_id,))
            logger.info(f"Deleted student: {student_id}")
            return True
        except Exception as e:
            logger.error(f"Error deleting student: {str(e)}")
            return False
    
    def find_all(self) -> List[Student]:
        """Find all students"""
        try:
            query = f"SELECT * FROM {self.keyspace}.students"
            result = self.session.execute(query)
            return [self._map_row_to_student(row) for row in result]
        except Exception as e:
            logger.error(f"Error finding all students: {str(e)}")
            return []
    
    def verify_password(self, student: Student, password: str) -> bool:
        """Verify student password"""
        return bcrypt.verify(password, student.password_hash)
    
    def _map_row_to_student(self, row) -> Student:
        """Map Cassandra row to Student model"""
        return Student(
            id=row.id,
            student_id=row.student_id,
            name=row.name,
            email=row.email,
            password_hash=row.password_hash,
            gpa=row.gpa,
            semester=row.semester,
            created_at=row.created_at,
            updated_at=row.updated_at
        )


# Singleton instance
student_repository = StudentRepository()
